<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CoPheScan: Input data • cophescan</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CoPheScan: Input data">
<meta property="og:description" content="cophescan">
<meta property="og:image" content="https://ichcha-m.github.io/cophescan/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cophescan</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/IntroductionCoPheScan_01.html">Introduction to CoPheScan</a>
    </li>
    <li>
      <a href="../articles/InputData_02.html">CoPheScan: Input data</a>
    </li>
    <li>
      <a href="../articles/FixedPriors_03.html">CoPheScan: Example with Fixed Priors</a>
    </li>
    <li>
      <a href="../articles/HierarchicalPriors_04.html">CoPheScan: Example with Hierarchical Priors</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ichcha-m/cophescan/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CoPheScan: Input data</h1>
                        <h4 data-toc-skip class="author">Ichcha
Manipur</h4>
            
            <h4 data-toc-skip class="date">2024-03-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ichcha-m/cophescan/blob/HEAD/vignettes/InputData_02.Rmd" class="external-link"><code>vignettes/InputData_02.Rmd</code></a></small>
      <div class="hidden name"><code>InputData_02.Rmd</code></div>

    </div>

    
    
<p>The input dataset for a trait (querytrait) should contain the summary
data for SNPs in a genomic region around the query variant (querysnpid)
and should have the following fields:</p>
<p><strong>For a Case-control dataset</strong></p>
<p>beta: <span class="math inline">\(\beta\)</span> or effect size</p>
<p>varbeta: variance of <span class="math inline">\(\beta\)</span> or
square of the standard error of <span class="math inline">\(\beta\)</span></p>
<p>snp: SNP identifier which maybe rsid or CHR_BP_REF_ALT or CHR_BP</p>
<p>type:‘cc’</p>
<p>N: sample size</p>
<p><strong>For a Quantitave dataset</strong></p>
<p>When, beta and varbeta are not available the following</p>
<p>beta: <span class="math inline">\(\beta\)</span> or effect size</p>
<p>varbeta: variance of <span class="math inline">\(\beta\)</span> or
square of the standard error of <span class="math inline">\(\beta\)</span></p>
<p>snp: SNP identifier which maybe rsid or CHR_BP_REF_ALT or CHR_BP</p>
<p>type:‘quant’</p>
<p>N: sample size</p>
<p>sdY: for a quantitative trait, the population standard deviation of
the trait.</p>
<p><strong>Additional fields in case of missing beta/varbeta or
sdY</strong></p>
<p>MAF: Minor allele frequency (only required when either beta/varbeta
or sdY are unavailable)</p>
<p>pvalues: only required when beta/varbeta are unavailable</p>
<p>s: fraction of samples that are cases (only for a case-control trait
when beta/varbeta are unavailable)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ichcha-m/cophescan" class="external-link">cophescan</a></span><span class="op">)</span></span></code></pre></div>
<p>Explore the data structure of the example dataset available in the
cophescan package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"cophe_multi_trait_data"</span><span class="op">)</span></span>
<span><span class="va">trait_dat</span> <span class="op">=</span> <span class="va">cophe_multi_trait_data</span><span class="op">$</span><span class="va">summ_stat</span><span class="op">$</span><span class="va">Trait_1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">trait_dat</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 8</span></span>
<span><span class="co">#&gt;  $ beta   : Named num [1:1000] -0.01369 0.01666 0.09057 -0.00571 -0.05606 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:1000] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span>
<span><span class="co">#&gt;  $ varbeta: Named num [1:1000] 0.000516 0.000399 0.003124 0.000419 0.000473 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:1000] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span>
<span><span class="co">#&gt;  $ z      : Named num [1:1000] -0.603 0.834 1.62 -0.279 -2.578 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:1000] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span>
<span><span class="co">#&gt;  $ snp    : chr [1:1000] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span>
<span><span class="co">#&gt;  $ MAF    : Named num [1:1000] 0.2614 0.4871 0.0318 0.4046 0.3042 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:1000] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span>
<span><span class="co">#&gt;  $ type   : chr "cc"</span></span>
<span><span class="co">#&gt;  $ N      : num 20000</span></span>
<span><span class="co">#&gt;  $ s      : num 0.5</span></span></code></pre></div>
<p><strong>Additional field for <code>cophe.susie</code></strong></p>
<p>LD: Linkage Disequilibrium matrix with row and column names being the
same as the snp field.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trait_dat</span><span class="op">$</span><span class="va">LD</span> <span class="op">=</span> <span class="va">cophe_multi_trait_data</span><span class="op">$</span><span class="va">LD</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">trait_dat</span><span class="op">$</span><span class="va">LD</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;  num [1:10, 1:10] 1 0.0267 -0.1078 -0.0627 0.1033 ...</span></span>
<span><span class="co">#&gt;  - attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:10] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:10] "chr19-11173352" "chr19-11173626" "chr19-11173716" "chr19-11173807" ...</span></span></code></pre></div>
<p>It is important to check that there is alignment of alleles for which
the beta is reported and those in the LD matrix. This can be verified
either using coloc::check_alignment or performing a diagnostic check
using the susie package <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link uri">https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html</a>.</p>
<p><strong>Note</strong></p>
<ul>
<li>The input summary data structure for a trait is the same format as
required by coloc. To gain a detailed understanding of the input please
see the coloc vignette <a href="https://chr1swallace.github.io/coloc/articles/a02_data.html" class="external-link uri">https://chr1swallace.github.io/coloc/articles/a02_data.html</a>
and the test datasets provided with the coloc package.</li>
<li>More information on the input can be obtained from the coloc
documentation : <code><a href="https://rdrr.io/pkg/coloc/man/check_dataset.html" class="external-link">?coloc::check_dataset</a></code>
</li>
</ul>
<hr>
</div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ichcha Manipur, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
