<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Bayesian cophescan analysis using Approximate Bayes Factors — cophe.single • cophescan</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Bayesian cophescan analysis using Approximate Bayes Factors — cophe.single"><meta property="og:description" content="Bayesian cophescan analysis under single causal variant assumption"><meta property="og:image" content="https://ichcha-m.github.io/cophescan/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cophescan</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/IntroductionCoPheScan_01.html">Introduction to CoPheScan</a>
    </li>
    <li>
      <a href="../articles/InputData_02.html">CoPheScan: Input data</a>
    </li>
    <li>
      <a href="../articles/FixedPriors_03.html">CoPheScan: Example with Fixed Priors</a>
    </li>
    <li>
      <a href="../articles/HierarchicalPriors_04.html">CoPheScan: Example with Hierarchical Priors</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ichcha-m/cophescan/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Bayesian cophescan analysis using Approximate Bayes Factors</h1>
    <small class="dont-index">Source: <a href="https://github.com/ichcha-m/cophescan/blob/HEAD/R/singlevar.R" class="external-link"><code>R/singlevar.R</code></a></small>
    <div class="hidden name"><code>cophe.single.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Bayesian cophescan analysis under single causal variant assumption</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cophe.single</span><span class="op">(</span></span>
<span>  <span class="va">dataset</span>,</span>
<span>  <span class="va">querysnpid</span>,</span>
<span>  <span class="va">querytrait</span>,</span>
<span>  MAF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pa <span class="op">=</span> <span class="fl">3.82e-05</span>,</span>
<span>  pc <span class="op">=</span> <span class="fl">0.00182</span>,</span>
<span>  p1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p12 <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dataset</dt>
<dd><p>a list with specifically named elements defining the query trait dataset
to be analysed.</p></dd>


<dt>querysnpid</dt>
<dd><p>Id of the query variant, (id in dataset$snp)</p></dd>


<dt>querytrait</dt>
<dd><p>Query trait name</p></dd>


<dt>MAF</dt>
<dd><p>Minor allele frequency vector</p></dd>


<dt>pa</dt>
<dd><p>prior probability that a non-query variant is causally associated with the query trait (cophescan prior), default 3.82e-5</p></dd>


<dt>pc</dt>
<dd><p>prior probability that the query variant is causally associated with the query trait (cophescan prior), default 1.82e-3 (cophescan prior)</p></dd>


<dt>p1</dt>
<dd><p>prior probability a SNP is associated with trait 1, (coloc prior), pc derived by using \(pc =  p12/p1+p12\); use p1, p2, p12 only when pa and pc are unavailable (See vignettes)</p></dd>


<dt>p2</dt>
<dd><p>prior probability a SNP is associated with trait 2,  (coloc prior), pa derived by using \(pa = p2\)</p></dd>


<dt>p12</dt>
<dd><p>prior probability a SNP is associated with both traits,  (coloc prior), pc derived by using \(pc =  p12/p1+p12\)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list of two <code>data.frame</code>s:</p><ul><li><p>summary is a vector giving the number of SNPs analysed, and the posterior probabilities of Hn (no shared causal variant), Ha (two distinct causal variants) and Hc (one common causal variant)</p></li>
<li><p>results is an annotated version of the input data containing log Approximate Bayes Factors and intermediate calculations, and the posterior probability SNP.PP.Hc of the SNP being causal for the shared signal <em>if</em> Hc is true. This is only relevant if the posterior support for Hc in summary is convincing.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function calculates posterior probabilities of different
causal variant configurations under the assumption of a single
causal variant for each trait.</p>
<p>If regression coefficients and variances are available, it
calculates Bayes factors for association at each SNP.  If only p
values are available, it uses an approximation that depends on the
SNP's MAF and ignores any uncertainty in imputation.  Regression
coefficients should be used if available. Find more input data structure details
in the coloc package</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Ichcha Manipur</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ichcha-m/cophescan" class="external-link">cophescan</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cophe_multi_trait_data</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">query_trait_1</span> <span class="op">&lt;-</span> <span class="va">cophe_multi_trait_data</span><span class="op">$</span><span class="va">summ_stat</span><span class="op">[[</span><span class="st">'Trait_1'</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">querysnpid</span> <span class="op">&lt;-</span> <span class="va">cophe_multi_trait_data</span><span class="op">$</span><span class="va">querysnpid</span></span></span>
<span class="r-in"><span><span class="va">res.single</span> <span class="op">&lt;-</span> <span class="fu">cophe.single</span><span class="op">(</span><span class="va">query_trait_1</span>, querysnpid <span class="op">=</span> <span class="va">querysnpid</span>, querytrait<span class="op">=</span><span class="st">'Trait_1'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running cophe.single...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP Priors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.96001823.82e-050.00182</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hypothesis Priors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.96001820.03816180.00182</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.002090.3750.623</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> PP for causal query variant: 62.3%</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">res.single</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nsnps      PP.Hn     PP.Ha     PP.Hc   lBF.Ha   lBF.Hc       querysnp</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  1000 0.00208927 0.3749638 0.6229469 15.32189 11.96576 chr19-11182353</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   querytrait typeBF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    Trait_1    ABF</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ichcha Manipur, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

